File: apps/relay-go/pkg/types/serialise.go
// Canonical state-hash computation per spec/serialisation.md §5.
//
//	stateHash = SHA-256( type_tag(0x0009) || canonical_encoding(state) )
//
// with state.state_hash itself encoded as 32 zero bytes inside the
// hashed input. The TypeScript reference at
// packages/protocol-types/src/serialise.ts MUST produce byte-identical
// output for byte-identical inputs — the cross-language vector at
// spec/test-vectors/state-hash.json is the binding surface.

package types

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"sort"
)

const (
	tagPlayer       uint16 = 0x0002
	tagRoundState   uint16 = 0x0009
	playerVersion   uint8  = 1
	roundStateVersion uint8 = 1
)

var zero32 = make([]byte, 32)

// ---------------------------------------------------------------------------
// Primitive encoders
// ---------------------------------------------------------------------------

func encU8(n uint8) []byte { return []byte{n} }

func encU16LE(n uint16) []byte {
	b := make([]byte, 2)
	binary.LittleEndian.PutUint16(b, n)
	return b
}

func encU32LE(n uint32) []byte {
	b := make([]byte, 4)
	binary.LittleEndian.PutUint32(b, n)
	return b
}

func encU64LE(n uint64) []byte {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, n)
	return b
}

func encBool(b bool) []byte {
	if b {
		return []byte{1}
	}
	return []byte{0}
}

func encVarint(n uint64) []byte {
	switch {
	case n < 0xfd:
		return []byte{byte(n)}
	case n <= 0xffff:
		out := make([]byte, 3)
		out[0] = 0xfd
		binary.LittleEndian.PutUint16(out[1:], uint16(n))
		return out
	case n <= 0xffffffff:
		out := make([]byte, 5)
		out[0] = 0xfe
		binary.LittleEndian.PutUint32(out[1:], uint32(n))
		return out
	default:
		out := make([]byte, 9)
		out[0] = 0xff
		binary.LittleEndian.PutUint64(out[1:], n)
		return out
	}
}

func encString(s string) []byte {
	bytes := []byte(s)
	return concat(encVarint(uint64(len(bytes))), bytes)
}

func encBytes32Hex(h string) ([]byte, error) {
	if len(h) != 64 {
		return nil, fmt.Errorf("encBytes32Hex: expected 64 hex chars, got %d (%q)", len(h), h)
	}
	out, err := hex.DecodeString(h)
	if err != nil {
		return nil, fmt.Errorf("encBytes32Hex: %w", err)
	}
	return out, nil
}

func encOptionalHash(h Hash256) ([]byte, error) {
	if h == "" {
		return encBool(false), nil
	}
	body, err := encBytes32Hex(string(h))
	if err != nil {
		return nil, err
	}
	return concat(encBool(true), body), nil
}

func encOptionalU32Ptr(n *BlockHeight) []byte {
	if n == nil {
		return encBool(false)
	}
	return concat(encBool(true), encU32LE(uint32(*n)))
}

func encOptionalSeatPtr(n *Seat) []byte {
	if n == nil {
		return encBool(false)
	}
	return concat(encBool(true), encU32LE(uint32(*n)))
}

func encOptionalString(s string) []byte {
	if s == "" {
		return encBool(false)
	}
	return concat(encBool(true), encString(s))
}

func encArrayBytes32(items []Hash256) ([]byte, error) {
	parts := [][]byte{encVarint(uint64(len(items)))}
	for _, h := range items {
		b, err := encBytes32Hex(string(h))
		if err != nil {
			return nil, err
		}
		parts = append(parts, b)
	}
	return concat(parts...), nil
}

func encArrayString(items []ActionType) []byte {
	parts := [][]byte{encVarint(uint64(len(items)))}
	for _, s := range items {
		parts = append(parts, encString(string(s)))
	}
	return concat(parts...)
}

func encArrayStringOutpoint(items []Outpoint) []byte {
	parts := [][]byte{encVarint(uint64(len(items)))}
	for _, s := range items {
		parts = append(parts, encString(string(s)))
	}
	return concat(parts...)
}

func encPreferences(prefs map[string]string) []byte {
	keys := make([]string, 0, len(prefs))
	for k := range prefs {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := [][]byte{encVarint(uint64(len(keys)))}
	for _, k := range keys {
		parts = append(parts, encString(k))
		parts = append(parts, encString(prefs[k]))
	}
	return concat(parts...)
}

func concat(parts ...[]byte) []byte {
	total := 0
	for _, p := range parts {
		total += len(p)
	}
	out := make([]byte, 0, total)
	for _, p := range parts {
		out = append(out, p...)
	}
	return out
}

// ---------------------------------------------------------------------------
// Composite encoders
// ---------------------------------------------------------------------------

func encPlayer(p PlayerState) ([]byte, error) {
	playerId, err := encBytes32Hex(string(p.PlayerId))
	if err != nil {
		return nil, fmt.Errorf("encPlayer: player_id: %w", err)
	}
	// Tail-32-of-pubkey, mirroring the TS encoder.
	pkHex := string(p.ValueSigningPubkey)
	if len(pkHex) < 64 {
		return nil, fmt.Errorf("encPlayer: pubkey shorter than 32 bytes hex: %q", pkHex)
	}
	pubkeyTail, err := encBytes32Hex(pkHex[len(pkHex)-64:])
	if err != nil {
		return nil, fmt.Errorf("encPlayer: pubkey tail: %w", err)
	}
	commitOpt, err := encOptionalHash(p.EntropyCommitmentHash)
	if err != nil {
		return nil, fmt.Errorf("encPlayer: entropy_commitment_hash: %w", err)
	}
	entropyOpt, err := encOptionalHash(p.EntropyValue)
	if err != nil {
		return nil, fmt.Errorf("encPlayer: entropy_value: %w", err)
	}
	return concat(
		encU16LE(tagPlayer),
		encU8(playerVersion),
		encU32LE(uint32(p.Seat)),
		playerId,
		pubkeyTail,
		encString(string(p.ParticipationStatus)),
		encU64LE(uint64(p.StakeAtRisk)),
		encOptionalString(string(p.StakeOutpoint)),
		encBool(p.EntropyCommitted),
		commitOpt,
		encBool(p.EntropyRevealed),
		entropyOpt,
		encArrayStringOutpoint(p.ConcealedCardRefs),
		encPreferences(p.DefaultPreferences),
	), nil
}

func encVisibleCards(cards []RevealedCard) []byte {
	parts := [][]byte{encVarint(uint64(len(cards)))}
	for _, c := range cards {
		parts = append(parts, encU8(uint8(c.Ordinal)))
	}
	return concat(parts...)
}

// EncodeRoundState returns the canonical encoding of a RoundState, with
// the state_hash field replaced by 32 zero bytes (self-reference is
// not part of the hashed input).
func EncodeRoundState(s RoundState) ([]byte, error) {
	playersParts := [][]byte{encVarint(uint64(len(s.Players)))}
	for _, p := range s.Players {
		enc, err := encPlayer(p)
		if err != nil {
			return nil, err
		}
		playersParts = append(playersParts, enc)
	}
	gameId, err := encBytes32Hex(string(s.GameId))
	if err != nil {
		return nil, fmt.Errorf("EncodeRoundState: game_id: %w", err)
	}
	ruleHash, err := encBytes32Hex(string(s.RuleSetHash))
	if err != nil {
		return nil, fmt.Errorf("EncodeRoundState: rule_set_hash: %w", err)
	}
	hiddenRefs, err := encArrayBytes32(s.HiddenCommitmentRefs)
	if err != nil {
		return nil, fmt.Errorf("EncodeRoundState: hidden_commitment_refs: %w", err)
	}
	successorRefs, err := encArrayBytes32(s.SuccessorTemplateHashes)
	if err != nil {
		return nil, fmt.Errorf("EncodeRoundState: successor_template_hashes: %w", err)
	}
	combinedOpt, err := encOptionalHash(s.CombinedEntropy)
	if err != nil {
		return nil, fmt.Errorf("EncodeRoundState: combined_entropy: %w", err)
	}
	deckOpt, err := encOptionalHash(s.DeckCommitmentHash)
	if err != nil {
		return nil, fmt.Errorf("EncodeRoundState: deck_commitment_hash: %w", err)
	}
	var priorOpt []byte
	if s.PriorStateHash != nil {
		priorOpt, err = encOptionalHash(*s.PriorStateHash)
		if err != nil {
			return nil, fmt.Errorf("EncodeRoundState: prior_state_hash: %w", err)
		}
	} else {
		priorOpt = encBool(false)
	}

	return concat(
		encU16LE(tagRoundState),
		encU8(roundStateVersion),
		encString(string(s.StateClass)),
		gameId,
		ruleHash,
		encU32LE(uint32(s.RoundNumber)),
		encOptionalSeatPtr(s.ActingPlayerSeat),
		concat(playersParts...),
		encU64LE(uint64(s.PotValue)),
		encVisibleCards(s.VisibleCards),
		hiddenRefs,
		encArrayString(s.AllowedActions),
		encOptionalU32Ptr(s.DecisionDeadlineBlockHeight),
		encOptionalU32Ptr(s.RecoveryDeadlineBlockHeight),
		successorRefs,
		combinedOpt,
		deckOpt,
		priorOpt,
		zero32,
	), nil
}

// ComputeStateHash returns the canonical state hash as a 64-char hex
// string.
func ComputeStateHash(s RoundState) (Hash256, error) {
	encoded, err := EncodeRoundState(s)
	if err != nil {
		return "", err
	}
	h := sha256.Sum256(encoded)
	return Hash256(hex.EncodeToString(h[:])), nil
}

// ChainsFromHash reports whether `next.prior_state_hash` equals
// ComputeStateHash(prior). The initial state of a session chains from
// nil.
func ChainsFromHash(prior *RoundState, next RoundState) (bool, error) {
	if prior == nil {
		return next.PriorStateHash == nil, nil
	}
	if next.PriorStateHash == nil {
		return false, nil
	}
	expected, err := ComputeStateHash(*prior)
	if err != nil {
		return false, err
	}
	return *next.PriorStateHash == expected, nil
}

File: packages/protocol-types/src/serialise.ts
/**
 * Canonical state-hash computation per `spec/serialisation.md` §5.
 *
 *   stateHash = SHA-256( type_tag(0x0009) || canonical_encoding(state) )
 *
 * with `state.state_hash` itself encoded as 32 zero bytes inside the
 * hashed input. Two honest implementations MUST produce byte-identical
 * outputs for the same logical input — the cross-language conformance
 * vector at `spec/test-vectors/state-hash.json` is the binding surface.
 *
 * Encoding rules (subset; full reference at `spec/serialisation.md`):
 *  - u8 / u16-LE / u32-LE / u64-LE: raw little-endian
 *  - varint:  Bitcoin-style (1, 3, 5, 9 bytes)
 *  - bytes32: 32 raw bytes
 *  - string:  varint(byte_len) || utf8 bytes (no NUL terminator)
 *  - bool:    1 byte (0 / 1)
 *  - optional<T>: bool(present) followed by T's encoding when present
 *  - array<T>:    varint(count) followed by each element in order
 *
 * Hashes are domain-separated by their type tag (`spec/serialisation.md` §4).
 */

import type { Hash256, RoundState } from './index.js';
import { asHash256 } from './primitives.js';

const TAG_PLAYER = 0x0002;
const TAG_ROUNDSTATE = 0x0009;
const ROUNDSTATE_VERSION = 1;
const PLAYER_VERSION = 1;

const ZERO_32 = new Uint8Array(32);

// -- primitive encoders ------------------------------------------------------

function encU8(n: number): Uint8Array {
  return new Uint8Array([n & 0xff]);
}
function encU16LE(n: number): Uint8Array {
  return new Uint8Array([n & 0xff, (n >>> 8) & 0xff]);
}
function encU32LE(n: number): Uint8Array {
  return new Uint8Array([n & 0xff, (n >>> 8) & 0xff, (n >>> 16) & 0xff, (n >>> 24) & 0xff]);
}
function encU64LE(n: number | bigint): Uint8Array {
  const v = typeof n === 'bigint' ? n : BigInt(n);
  const out = new Uint8Array(8);
  const dv = new DataView(out.buffer);
  dv.setBigUint64(0, v, true);
  return out;
}
function encBool(b: boolean): Uint8Array {
  return new Uint8Array([b ? 1 : 0]);
}
function encVarint(n: number): Uint8Array {
  if (n < 0xfd) return new Uint8Array([n]);
  if (n <= 0xffff) {
    const out = new Uint8Array(3);
    out[0] = 0xfd;
    out[1] = n & 0xff;
    out[2] = (n >>> 8) & 0xff;
    return out;
  }
  if (n <= 0xffffffff) {
    const out = new Uint8Array(5);
    out[0] = 0xfe;
    new DataView(out.buffer).setUint32(1, n >>> 0, true);
    return out;
  }
  const out = new Uint8Array(9);
  out[0] = 0xff;
  new DataView(out.buffer).setBigUint64(1, BigInt(n), true);
  return out;
}
function encString(s: string): Uint8Array {
  const bytes = new TextEncoder().encode(s);
  return concat(encVarint(bytes.byteLength), bytes);
}
function encBytes32Hex(h: Hash256 | string): Uint8Array {
  const hex = String(h);
  if (hex.length !== 64) throw new Error(`encBytes32Hex: expected 64 hex chars, got ${hex.length}`);
  const out = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    const byte = parseInt(hex.substring(i * 2, i * 2 + 2), 16);
    if (Number.isNaN(byte)) throw new Error(`encBytes32Hex: bad hex at ${i * 2}: ${hex}`);
    out[i] = byte;
  }
  return out;
}
function encOptionalHash(h: Hash256 | null): Uint8Array {
  return h === null ? encBool(false) : concat(encBool(true), encBytes32Hex(h));
}
function encOptionalU32(n: number | null): Uint8Array {
  return n === null ? encBool(false) : concat(encBool(true), encU32LE(n));
}
function encOptionalSeat(n: number | null): Uint8Array {
  return n === null ? encBool(false) : concat(encBool(true), encU32LE(n));
}
function encOptionalString(s: string | undefined): Uint8Array {
  return s === undefined || s === '' ? encBool(false) : concat(encBool(true), encString(s));
}
function encArrayBytes32(items: readonly (Hash256 | string)[]): Uint8Array {
  const parts: Uint8Array[] = [encVarint(items.length)];
  for (const h of items) parts.push(encBytes32Hex(h));
  return concat(...parts);
}
function encArrayString(items: readonly string[]): Uint8Array {
  const parts: Uint8Array[] = [encVarint(items.length)];
  for (const s of items) parts.push(encString(s));
  return concat(...parts);
}
function concat(...parts: readonly Uint8Array[]): Uint8Array {
  let total = 0;
  for (const p of parts) total += p.byteLength;
  const out = new Uint8Array(total);
  let off = 0;
  for (const p of parts) {
    out.set(p, off);
    off += p.byteLength;
  }
  return out;
}

// -- composite encoders ------------------------------------------------------

function encPlayer(p: RoundState['players'][number]): Uint8Array {
  return concat(
    encU16LE(TAG_PLAYER),
    encU8(PLAYER_VERSION),
    encU32LE(p.seat),
    encBytes32Hex(p.player_id),
    encBytes32Hex(p.value_signing_pubkey.substring(p.value_signing_pubkey.length - 64)), // tail 32 bytes
    encString(p.participation_status),
    encU64LE(p.stake_at_risk),
    encOptionalString(p.stake_outpoint),
    encBool(p.entropy_committed),
    encOptionalHash(p.entropy_commitment_hash),
    encBool(p.entropy_revealed),
    encOptionalHash(p.entropy_value),
    encArrayString([...p.concealed_card_refs]),
    encPreferences(p.default_preferences),
  );
}

function encPreferences(prefs: Readonly<Record<string, string>>): Uint8Array {
  const keys = Object.keys(prefs).sort(); // canonical comparator: lex byte order
  const parts: Uint8Array[] = [encVarint(keys.length)];
  for (const k of keys) {
    parts.push(encString(k));
    parts.push(encString(prefs[k] ?? ''));
  }
  return concat(...parts);
}

function encVisibleCards(cards: RoundState['visible_cards']): Uint8Array {
  const parts: Uint8Array[] = [encVarint(cards.length)];
  for (const c of cards) parts.push(encU8(c.ordinal));
  return concat(...parts);
}

/**
 * Canonical encoding of a RoundState, with the state_hash field
 * replaced by 32 zero bytes (so the hash can be embedded recursively
 * without circularity).
 */
export function encodeRoundState(state: RoundState): Uint8Array {
  const playersParts: Uint8Array[] = [encVarint(state.players.length)];
  for (const p of state.players) playersParts.push(encPlayer(p));
  return concat(
    encU16LE(TAG_ROUNDSTATE),
    encU8(ROUNDSTATE_VERSION),
    encString(state.state_class),
    encBytes32Hex(state.game_id),
    encBytes32Hex(state.rule_set_hash),
    encU32LE(state.round_number),
    encOptionalSeat(state.acting_player_seat),
    concat(...playersParts),
    encU64LE(state.pot_value),
    encVisibleCards(state.visible_cards),
    encArrayBytes32(state.hidden_commitment_refs),
    encArrayString([...state.allowed_actions]),
    encOptionalU32(state.decision_deadline_block_height),
    encOptionalU32(state.recovery_deadline_block_height),
    encArrayBytes32(state.successor_template_hashes),
    encOptionalHash(state.combined_entropy),
    encOptionalHash(state.deck_commitment_hash),
    encOptionalHash(state.prior_state_hash),
    ZERO_32, // state_hash placeholder (self-reference)
  );
}

/**
 * Compute the canonical state hash. Pure async (Web Crypto's
 * SubtleCrypto.digest is async).
 */
export async function computeStateHash(state: RoundState): Promise<Hash256> {
  const encoded = encodeRoundState(state);
  const buf = new ArrayBuffer(encoded.byteLength);
  new Uint8Array(buf).set(encoded);
  const digest = await crypto.subtle.digest('SHA-256', buf);
  return asHash256(toHex(new Uint8Array(digest)));
}

function toHex(b: Uint8Array): string {
  let s = '';
  for (let i = 0; i < b.byteLength; i++) {
    s += (b[i] as number).toString(16).padStart(2, '0');
  }
  return s;
}

/**
 * Verify that two consecutive states chain correctly: the second's
 * `prior_state_hash` MUST equal `computeStateHash(first)`.
 *
 * The initial state of a session chains from `null`.
 */
export async function chainsFromHash(
  prior: RoundState | null,
  next: RoundState,
): Promise<boolean> {
  if (prior === null) return next.prior_state_hash === null;
  if (next.prior_state_hash === null) return false;
  const expected = await computeStateHash(prior);
  return next.prior_state_hash === expected;
}
